source/device_defender.c

```diff
@@ -316,318 +565,832 @@ static uint64_t s_defender_report_id_epoch_time_ms(struct aws_iotdevice_defender
 
-            struct aws_byte_cursor report_topic = aws_byte_cursor_from_buf(&defender_task->report_topic_name);
-            struct aws_byte_cursor report = aws_byte_cursor_from_buf(&json_report);
+        if (defender_task->has_previous_net_xfer) {
+            delta_xfer.bytes_in = 0;
+            delta_xfer.bytes_out = 0;
+            delta_xfer.packets_in = 0;
+            delta_xfer.packets_out = 0;
 
-            AWS_LOGF_TRACE(
-                AWS_LS_IOTDEVICE_DEFENDER_TASK,
-                "id=%p: Full report: " PRInSTR,
-                (void *)defender_task,
-                AWS_BYTE_CURSOR_PRI(report));
-
-            uint16_t report_packet_id = aws_mqtt_client_connection_publish(
-                defender_task->config.connection,
-                &report_topic,
-                AWS_MQTT_QOS_AT_LEAST_ONCE,
-                false,
-                &report,
-                s_on_report_puback,
-                defender_task);
-
-            if (report_packet_id != 0) {
-                AWS_LOGF_DEBUG(
-                    AWS_LS_IOTDEVICE_DEFENDER_TASK,
-                    "id=%p: Report packet_id %d published on topic " PRInSTR,
-                    (void *)defender_task,
-                    report_packet_id,
-                    AWS_BYTE_CURSOR_PRI(report_topic));
-            } else {
-                AWS_LOGF_ERROR(
-                    AWS_LS_IOTDEVICE_DEFENDER_TASK,
-                    "id=%p: Report failed to publish on topic " PRInSTR,
-                    (void *)defender_task,
-                    AWS_BYTE_CURSOR_PRI(report_topic));
-            }
+            get_network_total_delta(&delta_xfer, &defender_task->previous_net_xfer, &totals);
+            ptr_delta_xfer = &delta_xfer;
 
-            for (size_t i = 0; i < net_conns.length; ++i) {
-                struct aws_iotdevice_metric_net_connection *con = NULL;
-                if (aws_array_list_get_at_ptr(&net_conns, (void **)&con, i)) {
-                    continue;
-                }
-                if (con->local_interface) {
-                    aws_string_destroy(con->local_interface);
-                }
-                if (con->remote_address) {
-                    aws_string_destroy(con->remote_address);
-                }
-            }
-            aws_array_list_clean_up(&net_conns);
-            aws_hash_table_clean_up(&ifconfig.iface_name_to_info);
+        } else {
+            defender_task->has_previous_net_xfer = true;
+        }
 
-            uint64_t now;
-            aws_event_loop_current_clock_time(defender_task->config.event_loop, &now);
-            aws_event_loop_schedule_task_future(
-                defender_task->config.event_loop, task, now + defender_task->config.task_period_ns);
+        /* serialize and publish MAY not publish successfully, but in the event
+           of failure, it will handle cleaning up the memory it allocated and
+           and invoking the task failure callback. If it succeeeds, completion
```

Comment:
I cannot parse the second sentence of this comment :)

Reply:
Reworded